fix: Prevent Legend Labels from Overlapping Diagram Elements in Journey Diagrams

Shahir-47 opened the pull request:
## :bookmark_tabs: Summary
Fixed an issue where long legend labels in Journey diagrams could overlap with the diagram elements. In addition, we've added a new journey theme variable, `maxLabelWidth`, which allows users to control the maximum width of actor labels. This variable is configured as follows:

```yaml
maxLabelWidth:
  description: Maximum width of actor labels
  type: integer
  default: 360
  minimum: 240
  maximum: 480
```

By tracking the maximum width of legend text elements, we now ensure proper spacing between the legend and the diagram content. This update also improves text-wrapping by:
- Breaking a long word with a hyphen if it exceeds the max width, then resuming the remaining word on the next line.
- Starting a new line without a hyphen if the break occurs at whitespace.

Resolves #5955

## :straight_ruler: Design Decisions
The implementation works by:
1. Tracking the maximum width of legend text elements during the actor legend drawing phase.
2. Using this width to calculate the appropriate left margin for the diagram elements.
3. Introducing the `maxLabelWidth` theme variable to let users set how long each label line should be.
4. Implementing word-wrapping behavior:
   - **Hyphenation:** If a word exceeds the line length, a hyphen is added at the break and the remaining text resumes on the next line.
   - **Whitespace Wrapping:** If the break occurs at a space, the new word starts on the next line without a hyphen.
5. Adding end-to-end tests to verify:
   - The proper wrapping behavior that adheres to the `maxLabelWidth` constraint.
   - That no line exceeds the specified max width, thereby maintaining consistent spacing between the legend and diagram.
   - The correct margin is maintained between the longest legend line and the diagram elements.

**Before**:

![image](https://github.com/user-attachments/assets/6f508a70-326c-4a92-9169-496bd3d99d96)

**After**:

![image](https://github.com/user-attachments/assets/00bf080a-8d99-4406-9bc8-2f945fc481a5)

### :clipboard: Tasks
- [x] :book: Have read the [[contribution guidelines](https://mermaid.js.org/community/contributing.html)](https://mermaid.js.org/community/contributing.html)
- [x] :computer: Have added necessary unit/e2e tests.
- [x] :notebook: Have added documentation. Make sure [`[MERMAID_RELEASE_VERSION](https://mermaid.js.org/community/contributing.html#update-documentation)`](https://mermaid.js.org/community/contributing.html#update-documentation) is used for all new features.
- [x] :butterfly: If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

Shahir-47 commented:
Hey @sidharthv96,

@pranavm2109 and I have fixed the issue with overlapping legend labels in Journey diagrams #5955. The fix ensures proper spacing by tracking the maximum width of legend text elements during the actor legend drawing phase and using this width to adjust the left margin for the diagram elements. We’ve also added E2E tests to verify that spacing remains correct for both normal and long labels.

All GitHub tests and CICD checks are passing, and this is ready to be merged. Full details are in the PR description—let us know if you have any questions or feedback!

sidharthv96 commented on packages/mermaid/src/diagrams/user-journey/journeyRenderer.ts:
```diff
@@ -84,31 +90,32 @@ export const draw = function (text, id, version, diagObj) {
 
   drawActorLegend(diagram);
-  bounds.insert(0, 0, LEFT_MARGIN, Object.keys(actors).length * 50);
+  leftMargin = conf.leftMargin + maxWidth - 22.328125;
```
What is `22.328125`, please avoid magic numbers.

pranavm2109 replied:
@sidharthv96 we added this number so that the spacing between the diagram and the legend text is consistent, otherwise the other e2e tests fail. Would it help to define the value 22.328125 as a constant?

sidharthv96 commented on cypress/integration/rendering/journey.spec.js:
```diff
@@ -63,4 +63,84 @@ section Checkout from website
+  });
+
+  it('should maintain sufficient space between legend and diagram when legend labels are longer', () => {
+    cy.then(() => {
+      renderGraph(
+        `journey
+      title  Web hook life cycle
+      section Darkoob
+        Make preBuilt:5: Darkoob user
+        register slug : 5: Darkoob userf deliberately increasing the size of this label to check if distance between legend and diagram is  maintained
+        Map slug to a Prebuilt Job:5: Darkoob user
+      section External Service
+        set Darkoob slug as hook for an Event : 5 : admin Exjjjnjjjj qwerty
+        listen to the events : 5 :  External Service
+        call darkoob endpoint : 5 : External Service
+      section Darkoob
+        check for inputs : 5 : DarkoobAPI
+        run the prebuilt job : 5 : DarkoobAPI
+        `,
+        { journey: { useMaxWidth: true } }
+      );
+    });
```
Why is the cy.then required, when it's not used in other functions?

sidharthv96 commented on packages/mermaid/src/diagrams/user-journey/journeyRenderer.ts:
```diff
@@ -39,14 +41,18 @@ function drawActorLegend(diagram) {
     };
-    svgDraw.drawText(diagram, labelData);
 
+    const textElement = svgDraw.drawText(diagram, labelData);
+    const textLength = textElement.node().getBBox().width;
+    if (textLength > maxWidth) {
+      maxWidth = textLength;
+    }
```
Not blocker: We should keep this dynamic width part, but should we split the text into more lines, if it's longer that a longer max limit?

Shahir-47 commented:
Hey @sidharthv96,

In this PR, @pranavm2109 and I introduced a new journey theme variable called `maxLabelWidth` to allow users to control the maximum width of actor labels in journey diagrams. The configuration is defined as follows:

```yaml
maxLabelWidth:
  description: Maximum width of actor labels
  type: integer
  default: 360
  minimum: 240
  maximum: 480
```

**Key updates include:**

- **Enhanced Text-Wrapping:**  
  - If a word exceeds the max width, it’s split with a hyphen at the break, and the remainder resumes on the next line.
  - If the break occurs at whitespace, the new word starts on a new line without any hyphen.

- **Improved Layout Consistency:**  
  - Added tests to ensure no line exceeds the `maxLabelWidth` so that the spacing between the label and diagram remains consistent.
  - Verified that when a label wraps, the appropriate margin is maintained between the longest legend line and the diagram.

- **Comprehensive Testing:**  
  - Tests confirm that a single long word wraps correctly with hyphenation.
  - Tests ensure that wrapping on whitespace omits the hyphen.
  - Tests also check that the spacing (margin) between the legend and the diagram is consistent based on the `maxLabelWidth` setting.

If there’s nothing else, this PR is good to go!

megantriplett commented:
Ran the deployment to the Mermaid live editor - and this fix works beautifully! Here it is with the original code:
![image](https://github.com/user-attachments/assets/3c0ccb43-9ca9-4244-ab48-abb409e6b5d1)
and when I add a lot of sections, the spacing remains clean for the users:
![image](https://github.com/user-attachments/assets/2bcce47e-7047-479a-be5f-6c210382ed67)
and even when I have a very long user name:
![image](https://github.com/user-attachments/assets/00a8ab36-be61-4cf0-b2cf-3d0d2a8cc9a7)
I think this is a good fix that should be merged!
